idb/postgres_migrations.go

```diff
@@ -0,0 +1,332 @@
+	for nextMigration < len(migrations) {
+		err = migrations[nextMigration](db, &state)
+		if err == asyncMigration {
+			// migration will continue asynchronously, it should call db.migrate() when it is done
+			return nil
+		}
+		if err != nil {
+			return fmt.Errorf("error in migration %d: %v", nextMigration, err)
+		}
+		nextMigration++
+	}
+	return nil
+}
+
+// after setting up a new database, mark state as if all migrations had been done
+func (db *PostgresIndexerDb) markMigrationsAsDone() (err error) {
+	state := MigrationState{
+		NextMigration: len(migrations),
+	}
+	migrationStateJson := string(json.Encode(state))
+	return db.SetMetastate(migrationMetastateKey, migrationStateJson)
+}
+
+func m0fixupTxid(db *PostgresIndexerDb, state *MigrationState) error {
+	mtxid := &txidFiuxpMigrationContext{db: db, state: state}
+	go mtxid.asyncTxidFixup()
+	return asyncMigration
+}
+
+func init() {
+	migrations = []migrationFunc{
+		m0fixupTxid,
+	}
+}
+
+const txidMigrationErrMsg = "ERROR migrating txns for txid, stopped, will retry on next indexer startup"
+
+type txidFiuxpMigrationContext struct {
+	db      *PostgresIndexerDb
+	state   *MigrationState
+	lastlog time.Time
+}
+
+// read batches of at least 2 blocks or up to 10000 txns,
+// write a temporary table, UPDATE from temporary table into txn.
+// repeat until all txns consumed.
+func (mtxid *txidFiuxpMigrationContext) asyncTxidFixup() {
+	db := mtxid.db
+	state := mtxid.state
+	log.Print("txid fixup migration starting")
+	prevRound := state.NextRound - 1
+	txns := db.YieldTxns(context.Background(), prevRound)
+	batch := make([]TxnRow, 15000)
+	txInBatch := 0
+	roundsInBatch := 0
+	prevBatchRound := uint64(math.MaxUint64)
+	var err error
+	for txr := range txns {
+		if txr.Error != nil {
+			log.Printf("ERROR migrating txns for txid rewrite: %v", txr.Error)
+			log.Print("txidMigrationErrMsg")
+			return
+		}
+		if txr.Round != prevBatchRound {
+			if txInBatch > 10000 {
+				err = mtxid.putTxidFixupBatch(batch[:txInBatch])
+				if err != nil {
+					return
+				}
+				// start next batch
+				batch[0] = txr
+				txInBatch = 1
+				roundsInBatch = 1
+				prevBatchRound = txr.Round
+				continue
+			}
+			roundsInBatch++
+			prevBatchRound = txr.Round
+		}
+		batch[txInBatch] = txr
+		txInBatch++
+		if roundsInBatch > 2 && txInBatch > 10000 {
+			// post the first complete rounds
+			split := txInBatch - 1
+			for batch[split].Round == txr.Round {
+				split--
+			}
+			split++ // now the first txn of the incomplete current round
+			err = mtxid.putTxidFixupBatch(batch[:split])
+			if err != nil {
+				return
+			}
+			// move incomplete round to next batch
+			copy(batch, batch[split:txInBatch])
+			txInBatch = txInBatch - split
+			roundsInBatch = 1
+			prevBatchRound = txr.Round
+			continue
+		}
+	}
+	if txInBatch > 0 {
+		err = mtxid.putTxidFixupBatch(batch[:txInBatch])
+		if err != nil {
+			return
+		}
+	}
+	// all done, mark migration state
+	state.NextMigration = 1
```

Comment:
Yeah, I see it, but also like the history of consensus protocol versions on mainnet, it's going to wind up an immutable chain of stuff and it will be a while before we can forget the old stuff. This migration will always be migration[0].